src/sage/rings/morphism.pyx

```diff
@@ -1109,13 +1109,99 @@ cdef class RingHomomorphism(RingMap):
             sage: psi.inverse_image(t^2) == z^2
             True
         """
+        from sage.rings.finite_rings.finite_field_base import FiniteField
+        from sage.rings.quotient_ring import QuotientRing_nc
+        if isinstance(self.domain(), QuotientRing_nc) and isinstance(self.codomain(), FiniteField):
+            if self.domain().base_ring().prime_subfield() == self.codomain().prime_subfield():
```

Comment:
```suggestion
            if self.domain().base_ring().prime_subfield() is self.codomain().prime_subfield():
```
Sage (especially algebraic) stuff usually inherit from `UniqueFactory` so `is` works and is faster (bypasses coercion).